python-checks/src/main/java/org/sonar/python/checks/PandasToDatetimeFormatCheck.java

```diff
@@ -0,0 +1,267 @@
+        reportIssue(ctx, yearfirstArgument, argumentExpression, expressionAndStringValue.originalExpression, String.format(MESSAGE, YEARFIRST, "False"));
+        return;
+      }
+    }
+  }
+
+  private static void reportIssue(SubscriptionContext ctx, RegularArgument dayfirstArgument, Expression argumentExpression, Expression originalExpression, String message) {
+    PreciseIssue preciseIssue = ctx.addIssue(dayfirstArgument, message);
+    if (argumentExpression != originalExpression) {
+      preciseIssue.secondary(argumentExpression, "This contains invalid date(s).");
+      preciseIssue.secondary(originalExpression, SECONDARY_MESSAGE);
+    } else {
+      preciseIssue.secondary(argumentExpression, SECONDARY_MESSAGE);
+    }
+  }
+
+  private static boolean getArgumentConstraint(@Nullable RegularArgument dayfirstArgument, Predicate<Expression> predicate) {
+    return Optional.ofNullable(dayfirstArgument)
+      .map(RegularArgument::expression).map(e -> {
+          Expression returnValue = e;
+          if (e.is(Tree.Kind.NAME) && isNotABooleanValue(e)) {
+            returnValue = Expressions.singleAssignedValue(((Name) e));
+          }
+          return returnValue;
+        }
+      ).map(predicate::test).orElse(false);
+  }
+
+  private static boolean isNotABooleanValue(Expression e) {
+    return !Expressions.isTruthy(e) && !Expressions.isFalsy(e);
+  }
+
+  private static List<ExpressionAndStringValue> getExpressionsAndStringValues(Expression expression) {
+    if (expression.is(Tree.Kind.STRING_LITERAL)) {
+      String normalizedStringValue = normalizeDateString(((StringLiteral) expression).trimmedQuotesValue());
+      return List.of(new ExpressionAndStringValue(expression, normalizedStringValue));
+    }
+    if (expression.is(Tree.Kind.NAME)) {
+      Optional<Expression> assignedValue = Expressions.singleAssignedNonNameValue((Name) expression);
+      return assignedValue.map(PandasToDatetimeFormatCheck::getExpressionsAndStringValues).orElse(Collections.emptyList());
+    }
+    if (expression.is(Tree.Kind.LIST_LITERAL)) {
+      return ((ListLiteral) expression).elements().expressions().stream()
+        .map(PandasToDatetimeFormatCheck::getExpressionsAndStringValues).flatMap(List::stream).collect(Collectors.toList());
+    }
+    return Collections.emptyList();
+  }
+
+  private static String normalizeDateString(String dateString) {
+    return dateString.trim().replaceAll("[\\./;\\s_]", "-");
+  }
+
+  private static ParseResult parseResult(String normalizedDateString, ParseResult expected) {
+    boolean parsedOnce = false;
+    for (Map.Entry<String, ParseResult> entry : DATE_FORMATS.entrySet()) {
+      if (expected.areAllConditionsMet()) {
+        return expected;
+      }
+      String dateFormat = entry.getKey();
+      ParseResult parseResult = entry.getValue();
+      DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern(dateFormat);
```

Comment:
I don't know about the performance impact, but it might be interesting to create all the DateTimeFormatter only once.